E2E: Introduce TestAccount / Refactor LoginPage

WunderBart opened the pull request:
This is a follow-up to https://github.com/Automattic/wp-calypso/pull/58516#discussion_r764400422.

### Suggested Changes

This PR is mainly about two things:

1. "Dumb down" the `LoginPage` model to remove (and prevent) the scope creep,
2. Introduce the `TestAccount` model to handle our e2e test accounts authentication.

In the spirit of keeping our `LoginPage` a simple POM, I came up with a concept of `TestAccount` model that would handle all the auth stuff related to our predefined E2E test accounts. In the first step, I created the new `LoginPage` from scratch, which was a great learning experience as it revealed, what I believe, some new good practices for our POMs - see the "Suggested good practices" section below. After many iterations of various account authentication concepts, I ended up creating the `TestAccount` model in the second step. It made sense to me to create this kind of high-level test account API as almost every suite (except account creation ones) revolves around our predefined test accounts. I think that from this place it would be best to let the code speak for itself, so please refer to the implementation for more details! 🙏 

Other changes:
- Make it possible to pass test account names via the `SAVE_AUTH_COOKIES` env var.
  This helps quickly pre-authenticate selected account(s) for faster local testing.
- Make the `setupHook` callback function async.
  This way it can be used also as `beforeAll` hook so we don't need to add an extra one.
- Remove the explicit `Log in` step and authenticate via `TestAccount` in the `setupHook`.
  We're authenticating some accounts via saved cookies, which is not a log-in step per se.
- Remove `clearAuthenticationState()` and `setLoginCookie()` from `BrowserManager`.
  This logic has been moved to the `TestAccount` model.
- Remove `startImmediately()` from the `NewPostFlow` as it's no longer used.
  As above, it's now handled by `TestAccount`. 
- Implement and use `gutenbergEditorPage.visit( 'post' )` instead of `newPostFlow.newPostFromNavbar()` to get to the "new post" page.
  See the "Define `visit()` method for each Page model" good practice suggestion below for the rationale on this one.
- Some general clean up.

### Testing instructions

- All tests should pass.
- Make sure default common accounts defined in the `globalSetup.ts` are pre-authenticated when the `SAVE_AUTH_COOKIES` var is set to `true`.
- Make sure pre-authentication is turned off when `SAVE_AUTH_COOKIES` var is set to `false`. 
- Make sure custom accounts are pre-authenticated when passed via the `SAVE_AUTH_COOKIES`, e.g.

  ```sh
  > export SAVE_AUTH_COOKIES=p2User,i18nUser
  > npx jest specs/specs-p2/p2-post.ts
  ```
- Make sure the authentication information is printed out when the `DEBUG` var is defined, e.g.

  ```sh
  > export DEBUG=pw:api
  > npx jest specs-p2/p2-post.ts
  ```
- Make sure the auth cookies are correctly considered `fresh` and `stale`.
  To test this one you can alter the 3 days value to just a few seconds in the `TestAccount.hasFreshCookies` method.

### Next steps

I've noticed that in some tests we use multiple accounts for the same browser context, for example [wp-support__showme-where.ts](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/specs/specs-playwright/wp-support__showme-where.ts). This forces us to manually purge the current context in order to provide a fresh login state. The biggest disadvantage of this is that we cannot be sure we have a 100% separation this way, which we should. I think that with a fairly simple change we can create a new context for each `describe` block, which would let us free from worrying about any cookie leftovers, local states, or sessions. This would probably be good as a separate issue though - will create one!

---

### Suggested good practices for Page Model Objects based on the refactored LoginPage:

- **Define `visit()` method for each Page model.** 
  This helps quickly identify which page is handled by the current model. It also helps get directly to that page when needed instead of using flows to click the user through the navigation UI. I think that using navigation flows as a way of getting to the target page creates a risk of failure before even getting to the actual test subject. For example, we have a lot of editor test suites, each of which uses the Calypso interface to navigate to the editor page. Opening the editor via the UI should be tested separately, and for the actual editor tests, we can go to its page directly saving precious time and increasing test stability. As an example, this PR introduces also the `gutenbergEditorPage.visit()` method.
- **Don't redirect within page methods.**
  In other words, the redirect to the target page should always be called explicitly. Sometimes, the same page can have different URLs (like e.g. the log-in popup), and every POM method should be available for immediate use when the page is rendered, regardless of the current URL.
  As a supporting example, the actual log-in page sometimes opens as a popup window. If any of the LoginPage methods implied redirect (as e.g. `loginPage.logIn()` previously did), it would not be usable on the popped-up page window.  
- **Create small, single element/responsibility user actions.** 
  It makes the creation of the `selectors` object redundant and helps write readable flows where each step is self-described, e.g.
  
  ```js
  class LoginPage {
  // ...
  	async fillUsername( value: string ): Promise< Locator > {
  		const locator = await this.page.locator( 'input#usernameOrEmail' );
  		await locator.fill( value );
  	
  		return locator;
 	 }

  	async clickSubmit(): Promise< Locator > {
  		const locator = await this.page.locator( 'button[type="submit"]' );
  		await locator.click();

  		return locator;
  	}
  // etc.
  }
  ```
- **Use only existing element actions to create batch actions.**

  For example:

  ```js
  class LoginPage {
  // ...
  	async logIn( username, password ) {
  		await this.fillUsername( 'admin' );
  		await this.fillPassword( 'password' );
  		await Promise.all( [
  			this.page.waitForNavigation(),
  			this.clickSubmit(),
  		] );
  	}
  // etc.
  }
  ```

WunderBart committed:
Initialize LoginPage

WunderBart committed:
Add logInWithUsername to LoginPage

WunderBart committed:
Add LoginPage.loginWithTestAccount

WunderBart committed:
Refactor login step for CoBlocks tests

WunderBart committed:
Refactor login step for wp-reader__view-spec.ts

WunderBart committed:
Update login step for remaining tests

WunderBart committed:
Make testAccount var naming consistent

WunderBart committed:
Fix wp-support__showme-where.ts

WunderBart committed:
Fix tests where account is switched

WunderBart committed:
Fix wp-user__notifications.ts

WunderBart committed:
Fix wp-invite__new-user.ts

WunderBart committed:
Streamline logging in

WunderBart committed:
Stabilize wp-user__notifications specs

WunderBart committed:
Streamline LoginPage even more

WunderBart committed:
Create authenticateTestAccount method

WunderBart committed:
Clear cookies prior to authentication

WunderBart committed:
Promisify setupHooks callback fn

WunderBart committed:
Wait for added block only to be attached

WunderBart committed:
Address #54421 once again

WunderBart committed:
Forgot seconds :]

WunderBart committed:
Force dismiss the Welcome Tour (see #57660)

WunderBart committed:
E2E: Forcefully close the Welcome Tour popup

WunderBart committed:
Make sure the Tour status is loaded

WunderBart committed:
Update after rebasing with fix/e2e-force-close-welcome-tour

WunderBart committed:
Extract the cookie saving part to globalSetup

WunderBart committed:
Update CoBlocks suites

WunderBart committed:
Create TestAccount model

WunderBart committed:
Use TestAccount model in some suites

WunderBart committed:
Use good ole array for creds

WunderBart committed:
Do some cleaning up

WunderBart committed:
Use TestAccount in wp-reader__view-spec.ts

WunderBart committed:
TestAccount: login with TOTP if available

WunderBart committed:
Use TestAccount in wp-i18n__editor.ts

WunderBart committed:
Use TestAccount in p2-post.ts suite

WunderBart committed:
Use TestAccount in wp-blocks__coblocks__extensions__replace-image.ts suite

WunderBart committed:
Use TestAccount in wp-domains__add.ts suite

WunderBart committed:
Make account siteURL a getter so it throws when the URL isn't available

WunderBart committed:
Use TestAccount in remaining suites

WunderBart committed:
Fix wp-reader__view-spec.ts

WunderBart committed:
Do not import BorwserManager unnecessarily

WunderBart committed:
See if we can live without the r-login clearance

WunderBart committed:
Remove obsolete auth clearing step

WunderBart committed:
Add docs

WunderBart committed:
Revert unnecessary changes

WunderBart committed:
Fix step where cookies are not re-created when saved

WunderBart committed:
Validate env vars better

worldomonation commented:
I only realized now that I've been tagged in this review; sorry about that.

I will allocate some time to review in the next two days and return with some feedback.

worldomonation commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -105,7 +105,7 @@ export function getCalypsoURL(
  * @throws {Error} If accountType does not correspond to a valid entry in the file.
  */
-export function getAccountCredential( accountType: string ): string[] {
+export function getAccountCredential( accountType: string ): [ string, string ] {
```
I assume this change is to use the correct signature? I don't see anything else having changed in this method.

WunderBart replied:
This narrows the returned type so that it's spreadable, e.g.:

```js
await logIn( ...getAccountCredential() ); // where the logIn method accepts exactly two arguments
```

worldomonation commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -74,6 +70,19 @@ export class GutenbergEditorPage {
+	 * Opens the "new post/page" page. By default it will open the "new post" page.
+	 *
+	 * Example "new post": {@link https://wordpress.com/post}
+	 * Example "new page": {@link https://wordpress.com/page}
```
nit: missing JSDoc on the return value.

worldomonation commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -86,6 +95,8 @@ export class GutenbergEditorPage {
 		// takes too long here, so the most reliable alternative is the title being
 		// visible.
 		await frame.waitForSelector( selectors.editorTitle );
+		// Once https://github.com/Automattic/wp-calypso/issues/57660 is resolved,
+		// the next line should be removed.
```
Not necessarily related to this PR, though I think even after #57660 is fixed this should remain in case.

worldomonation commented on packages/calypso-e2e/src/lib/test-account.ts:
```diff
@@ -0,0 +1,161 @@
+		await loginPage.visit();
+		await loginPage.logInWithCredentials( ...this.credentials );
+
+		const verificationCode = this.getTOTP();
```
Clever!

_Eventually_ I want to do a clean sweep of the config file, since it's full of cruft and unused things at this point, and establish a much more robust data structure to hold the username, password, TOTP secret and other features of the account. It would look similar to the current config file's `accounts` entry.

worldomonation commented on packages/calypso-e2e/src/lib/test-account.ts:
```diff
@@ -0,0 +1,161 @@
+		try {
+			const { birthtimeMs } = await fs.stat( this.getAuthCookiesPath() );
+			const nowMs = Date.now();
+			const threeDaysMs = 3 * 24 * 60 * 60 * 1000;
```
I'm getting conflicting information on this one, maybe you can chime in with your experience.

I've had cookies be rejected at both 1 and 3 days. To be safe, I think it might be good idea to have the expiry date set to 1 day, but I'd like to base any decisions off an actual value as implemented in the backend if possible.

WunderBart replied:
Hmm, I wasn't able to test that on my end (can't seem to get the cookies rejected), but this might need some deeper look. I'd move that to a follow-up PR. To be clear, are you able to pre-authenticate with freshly created cookies though?

worldomonation replied:
> Hmm, I wasn't able to test that on my end (can't seem to get the cookies rejected), but this might need some deeper look. I'd move that to a follow-up PR. To be clear, are you able to pre-authenticate with freshly created cookies though?

To clarify, this was an observation of mine dating back to the early days of implementing pre-authenticated cookies. I haven't tested your changes locally, but have no doubts that it works! No changes are required for this PR.

This most recently came up in December (prior to my AFK) when I was looking at preserving cookies for the P2 E2E user to work around [this issue](https://github.com/Automattic/wp-calypso/issues/59211), but I kept having my cookies rejected after 3 days at most. This is despite the P2 team apparently successfully using a cookie they had preserved many months ago. 

To add to the confusion (for me at least) the cookie expiry date is set for many months ahead, not just 3 days. Going off the cookie expiry timestamp, there is no reason for it (cookie) to be expired after merely 3 days.

worldomonation replied:
I ran into rejected cookies despite it being within the 3 day period.

Cookie file I had locally (times in PST):
`10 Jan 15:03  simpleSitePersonalPlanUser.json`

Calypso rejected the cookie I had however, throwing me to the login screen.

The `wordpress_logged_in_525ed659f` entry had an expiry timestamp of 1644447823.194031, which works out to 02-09-2022 15:03:43 PST. 

So, something is fishy or not working as intended here. In the meantime, I suspect it might be better to reduce the cookie freshness threshold to 2 days or 2 * 24 * 60 * 60 * 1000.

worldomonation commented on packages/calypso-e2e/src/lib/test-account.ts:
```diff
@@ -0,0 +1,161 @@
+	 * is defined. Prefixes each message with the account name for easier
+	 * reference. Formatted similarly to the pw:api logs.
+	 */
+	private log( message: any ) {
+		const { DEBUG } = process.env;
+		if ( DEBUG && DEBUG !== 'false' ) {
+			console.log(
+				`${ chalk.bold( chalk.magenta( `TestAccount:${ this.accountName }` ) ) } => ${ message }`
+			);
```
👍

worldomonation commented on test/e2e/specs/specs-playwright/wp-likes__post-spec.js:
```diff
@@ -66,26 +62,21 @@ describe( DataHelper.createSuiteTitle( 'Likes (Post)' ), function () {
 			publishedPostPage = new PublishedPostPage( page );
-
-			// Clicking the Like button will bring up a new popup.
-			// Observe for the new popup in the Promise. Once the Page object
-			// is obtained, execute login steps there and wait for the
-			// popup to close.
-			await Promise.all( [
-				page.on( 'popup', async ( popup ) => {
-					const loginPage = new LoginPage( popup );
-					await loginPage.loginFromPopup( { account: likeUser } );
-					await popup.waitForEvent( 'close' );
-				} ),
-				publishedPostPage.likePost(),
-			] );
+			await publishedPostPage.likePost();
```
Does this change preserve the steps of:
- user is logged out
- user clicks on Like Post button
- user is prompted to log in
- once logged in, Like is recorded

WunderBart replied:
Good point, I missed the fact that the popup itself might be a subject of the test here. Will update!

WunderBart replied:
I'll handle that in a follow-up PR where we can separate the context properly for both, posting and liking users: #59811.

WunderBart replied:
https://github.com/Automattic/wp-calypso/pull/59811/commits/281250c606b838029a870185cbc36f9374ef811d

worldomonation reviewed (approved):
Bravo 🥂 

I think you have done a great job here. The description was clear and easy to follow along and the code is just as easy to follow along. There were some tricky parts I would've liked a comment on, but those are minor nits.

In particular, I think the move to abstract the actual login logic to `TestAccounts` makes a lot of sense.

Approving to unblock your progress. Most comments I left were just that, comments or questions. Feel free to address them or not, depending on how much time you have left.

worldomonation commented:
> Don't redirect within page methods.
In other words, the redirect to the target page should always be called explicitly. Sometimes, the same page can have different URLs (like e.g. the log-in popup), and every POM method should be available for immediate use when the page is rendered, regardless of the current URL.
As a supporting example, the actual log-in page sometimes opens as a popup window. If any of the LoginPage methods implied redirect (as e.g. loginPage.logIn() previously did), it would not be usable on the popped-up page window.

Could you elaborate on this if you don't mind? I would like to better understand what you mean by 'redirect within methods'.

worldomonation commented:
> Create small, single element/responsibility user actions.
It makes the creation of the selectors object redundant and helps write readable flows where each step is self-described, e.g.

Good idea. I still see a place for the `selector` object, especially for dynamic selectors or selectors found across multiple methods, but I feel comfortable incorporating this recommendation into the style guide such that:

- if selector is used only once, no need to put into object.
- if selector is used twice or more, use an object.

WunderBart committed:
Merge trunk

WunderBart commented:
> > Create small, single element/responsibility user actions.
> > It makes the creation of the selectors object redundant and helps write readable flows where each step is self-described, e.g.
> 
> Good idea. I still see a place for the `selector` object, especially for dynamic selectors or selectors found across multiple methods, but I feel comfortable incorporating this recommendation into the style guide such that:
> 
> * if selector is used only once, no need to put into object.
> * if selector is used twice or more, use an object.

I think that with the `locator` API we might not need the `selectors` object at all. We can initialize all the elements **in the constructor** and reference them in the POM and the tests as well. There's a great short guide on this here https://playwright.dev/docs/test-pom. Here's an example from that page:

```js
// playwright-dev-page.js
const { expect } = require('@playwright/test');

exports.PlaywrightDevPage = class PlaywrightDevPage {

  /**
   * @param {import('@playwright/test').Page} page
   */
  constructor(page) {
    this.page = page;
    this.getStartedLink = page.locator('text=Get started');
    this.coreConceptsLink = page.locator('text=Core concepts');
    this.tocList = page.locator('article ul > li > a');
  }

  async goto() {
    await this.page.goto('https://playwright.dev');
  }

  async getStarted() {
    await this.getStartedLink.first().click();
    await expect(this.coreConceptsLink).toBeVisible();
  }

  async coreConcepts() {
    await this.getStarted();
    await this.page.click('text=Guides');
    await this.coreConceptsLink.click();
    await expect(this.page.locator('h1').locator("text=Core concepts")).toBeVisible();
  }
}
```

WunderBart commented:
> > Don't redirect within page methods.
> > In other words, the redirect to the target page should always be called explicitly. Sometimes, the same page can have different URLs (like e.g. the log-in popup), and every POM method should be available for immediate use when the page is rendered, regardless of the current URL.
> > As a supporting example, the actual log-in page sometimes opens as a popup window. If any of the LoginPage methods implied redirect (as e.g. loginPage.logIn() previously did), it would not be usable on the popped-up page window.
> 
> Could you elaborate on this if you don't mind? I would like to better understand what you mean by 'redirect within methods'.

I guess what I'm trying to say, is that the POM methods that represent user actions should not make a `goto` call within them, as it's **not a user action**. It also introduces a bad practice where some methods suddenly redirect without a clear reason (e.g. which methods should redirect to the target page and which shouldn't?). As a practical example: 

**`BAD`**
```js
/* pages/media-page.js */

class MediaPage {
  async visit() {
    return await this.page.goto('https://wordpress.com/media');
  }
  await addMedia( files ) {
    await this.visit();
    await this.addMediaButton.click();
    await this.mediaInput.setInputFiles( files );
  }
}

/* tests/media-spec.js */

it('Do something on some other page', async () => {
  // clicking and expecting stuff
});

it('Add some photos', async () => {
  await readerPage.addMedia( testImage ); // Silent redirect; How did the user get there?
});
```

**`GOOD`**
```js
/* pages/media-page.js */

class MediaPage {
  async visit() {
    return await this.page.goto('https://wordpress.com/media');
  }
  await addMedia() {
    await this.addMediaButton.click();
  }
}

/* tests/media-spec.js */

it('Do something on some other page', async () => {
  // clicking and expecting stuff
});

it('Go to the Media page', async () => {
  await readerPage.visit(); // Explicit redirect; All clear now.
});

it('Add some photos', async () => {
  await readerPage.addMedia( testImage ); 
});
```

I'm not sure if it's the best example but hopefully, it clears the concept up a bit! 😅

WunderBart committed:
Merge branch 'trunk' into refactor/e2e-login-flow

WunderBart committed:
Fix that aren't failing (but should?)

I'm not sure why e.g. the signup and magic link specs didn't run on this PR by default. My changes did break them.

WunderBart committed:
Fix crowdsignal signup spec

WunderBart commented:
[That one Gutenberg spec that failed](https://teamcity.a8c.com/viewLog.html?buildId=7243222&buildTypeId=calypso_WPComTests_gutenberg_desktop) is unrelated to current changes. Merging!
